### ISSUE TITLE
Add support for a whitespace removal block

### ISSUE THREAD

**Issue by cipriancraciun** [djc/askama#317]
Go's [quicktemplate](https://github.com/valyala/quicktemplate) has a nice feature that opens a block in which all template statements behave like they would have been prefixed with `{%-` and `-%}`.  (See the example bellow.)

This is useful especially when whitespace is important in the resulting output (and we want to eliminate it), but we still want to use it in our templates for readability.  (For example XML.)

It would be nice to have such a `{% stripspace %} ... {% endstripspace %}` to achieve that.  (Else at the moment the user has to carefully put the `-` in all his tags.)

~~~~
{% stripspace %}
     <div>
         <div>space between lines</div>
            and {%s " tags" %}
         <div>is removed unless{% newline %}or{% space %}is used</div>
     </div>
{% endstripspace %}
~~~~

Note this feature wouldn't remove any other whitespace that a plain `{%- .. -%}` wouldn't remove by itself.  For example `abc cde` would be rendered exactly as that, but `abc {{ "cde" }} efg` would be rendered as `abccdeefg`.

**Comment by cipriancraciun** [comment]
I forgot to mention that in case of `quicktemplate` two helper items were added `{% newline %}` and `{% space %}` to make it easier to include such whitespace elements in the context of `{% stripspace %}`.

**Comment by cipriancraciun** [comment]
A good use-case for such a feature is generating plain-text files with careful formatting, where the easiest way to achieve this is to wrap the whole template in a `{% stripspace %}` construct and then anything that is to be output should be controlled by `{{ "whatever" }}`.

**Comment by djc** [comment]
Yeah, I suppose that could be useful sometimes.

**Comment by cipriancraciun** [comment]
@djc OK, I'll try to prepare a pull request in the weekend.

But before that there is a detail regarding nesting of this block and the other blocks.  For example the following would be useful:
~~~~
{% stripspace %}
    {% for i in 0..100 %}
        ... some code where we don't want white space
        {% endstripspace %}
        .. some code where we do want verbatim white space
        {% stripspace %}
        ... again some code without white space
    {% endfor %}
{% endstripspace %}
~~~~

In the example above `stripspace` doesn't respect the nesting of `for`.  It breaks the rule of "strict" nesting of block constructs.

However at the same time, `stripspace` isn't an actual block item, or flow control item, but instead it changes the way the parser works.

Therefore the question is:  should strict nesting be enforced?

Also, should I add `{{ space }}` and `{{ newline }}` (and perhaps `{{ tab }}`) as additional items?

**Comment by djc** [comment]
Yes, strict nesting should be enforced.

I don't think we need `{{ space }}` (you mentioned `{% space %}` before -- which is it?) et al. Seems to me like `{{ ' ' }}` is actually shorter and more obvious.

**Comment by cipriancraciun** [comment]
> I don't think we need {{ space }} (you mentioned {% space %} before -- which is it?)

Yes, sorry, it should have been `{% space %}`.

The main issue with `{{ ' ' }}` is that it can't be find-and-replace easiy, meanwhile `{%space%}` can.  (The same for `{%newline%}` and `{%tab%}`.)

**Comment by djc** [comment]
I don't follow why finding and replacing that is not easy?

**Comment by cipriancraciun** [comment]
> I don't follow why finding and replacing that is not easy?

For example with `git grep` it takes quite a few escapes to get `'` or `"` (because the user can choose to use either `{{ " " }}` or `{{ ' ' }}`.

This is even worse for grepping for `{{ "\n" }}`.

Basically by using `{%space%}` it makes it very easy to spot.

**Comment by djc** [comment]
I don't think that use case is important enough to introduce the special-case abstraction.

**Comment by cipriancraciun** [comment]
OK, I've made my proof-of-concept for the proposed feature;  the diff can be seen bellow:
https://github.com/cipriancraciun/askama/compare/patches/upstream/stripspace--base...cipriancraciun:patches/upstream/stripspace

(The diff above contains both `stripspace` addition as a single commit, and also the proposed `{% space %}`, `{% newline %}` and `{% tab %}` additions as another commit.)

----

Now about the `{% space %}` and friends.  Besides aiding in the `grep`-ing, they improve performance and simplify the generated code as seen from the following example.

Given  the template `{% space %}{% tab %}{% newline %}`, the generated code is:
~~~~
    fn render_into(&self, writer: &mut ::std::fmt::Write) -> ::askama::Result<()> {
        writer.write_str(" \t\n")?;
        Ok(())
    }
~~~~

Given the template `{{ " " }}{{ "\t" }}{{ "\n" }}`, the generated code is:
~~~~
    fn render_into(&self, writer: &mut ::std::fmt::Write) -> ::askama::Result<()> {
        write!(
            writer,
            "{expr0}{expr1}{expr2}",
            expr0 = &::askama::MarkupDisplay::new_unsafe(&" ", ::askama::Text),
            expr1 = &::askama::MarkupDisplay::new_unsafe(&"\t", ::askama::Text),
            expr2 = &::askama::MarkupDisplay::new_unsafe(&"\n", ::askama::Text),
        )?; 
        Ok(())
    }   
~~~~

Now there is a subtle difference between `{% space %}` and `{{ " " }}`, in that the former bypasses the escaping and behaves just like a literal would have, which is exactly what is needed when using `{% stripspace %}` contexts.

As such, without the addition of these special markers, the `{% stripspace %}` feature isn't 100% complete.

----

@djc any comments on the proposed patch before submitting a pull request?

**Comment by djc** [comment]
I still don't want the special case `space` et al blocks. However, it'd be good to improve the code generation for string literals to be smarter such that all string literals can enjoy generated code like your proposed new blocks would.

**Comment by cipriancraciun** [comment]
> However, it'd be good to improve the code generation for string literals to be smarter such that all string literals can enjoy generated code like your proposed new blocks would.

This improvement is not doable because semantically the two expressions (i.e. `{% space %}` and `{{ " " }}`) are different:
* the first one is an actual "literal";  (I.e. `1 2` is equivalent to `1{% space %}2`;)
* the second one is an expression that will pass through the escaper;

Thus, unless we want for `{{ }}` to sometimes be escaped and sometimes not depending on the actual value, it can't be done.

Or to put it another way, assume one has a special output format where spaces are not allowed and must be escaped.  If we make `{{ " " }}` a special case, then these won't pass through the escaper and will be broken.

Therefore `{% space %}` and friends are actual literals that don't pass through the escaper.

**Comment by djc** [comment]
That doesn't make sense to me. String literals are literals and therefore known at compile time. Therefore we can treat them exactly the same as some hypothetical syntax construct that evaluates to some particular string.

**Comment by cipriancraciun** [comment]
> String literals are literals and therefore known at compile time.

Yes, but those string literals are inside "expression" blocks, i.e. `{{ "x" }}` which is different from a string literal outside an expression block (i.e. plain `x`).

To put things differently, the following two templates are not rendered the same if one switches the escaper to something that replaces say `x` with `Y`:
* `x` would yield `x`;
* `{{ "x" }}` would field `Y`;

Now apply the same reasoning for spaces:
* ` ` (i.e. a single space) would always yield ` ` regardless of the escaper;
* `{% space %}` would be semantically equivalent to the previous one, thus always yielding ` `;
* `{{ " " }}`, if used with an escaper that treats spaces specially would yield something else;

----

Concrete example:  say I want to generate `bash` scripts, where the user input (part of the template fields) are to be used as command arguments.  In that case I would use an escaper that will always quote strings that contain special characters.  Therefore if I would also want to use `{% stripspace %}` it would be impossible for me now to actually output an unescaped space.

**Comment by djc** [comment]
It's a string literal, so we can evaluate the escaped value at compile time rather than at run time.

**Comment by cipriancraciun** [comment]
> It's a string literal, so we can evaluate the escaped value at compile time rather than at run time.

OK, but then how do I output a space that doesn't pass through the escaper?  I use the "safe" filter?  So now basically instead of `{% space %}` I have to write `{{ " " | safe }}`.

(Furthermore I don't think you can / should evaluate it at compile time because the implementation of the escaper should run at run-time, else you'll have some parts of the escaper run at compile time and some at run time.  Moreover that would mean that the escaper should be build-able outside of the crate that contains the template.)

----

Anyway, I think I've spent quite some time supporting the `{% space %}` and friends.  I'll submit a pull request with both `{% stripspace %}` and `{% space %}` and friends.  If it's not accepted I'll have to fork the library and maintain the patch myself, because as said `{% stripspace %}` is not "complete" without the other three...  :(

**Comment by djc** [comment]
Outputting a space that doesn't pass through the escaper wouldn't matter anymore, because the escaper would do the right thing at compile time, and so escaping would be without run time costs. I don't see any problems with running escaping for compile time-known data at compile time and running the rest at run time. Yes, the escaper should be buildable but I don't see how that's a problem, either.

Forking is fine of course. It doesn't seem like an optimal solution to me, but it is your prerogative.

**Comment by cipriancraciun** [comment]
> Yes, the escaper should be buildable but I don't see how that's a problem, either.

Because the user would have to put the escaper in a different crate, complicating further the usage.

**Comment by djc** [comment]
This is true. However, I feel like this would still be true for the hypothetical case of an escaper that wants to potentially escape things like `space` et al even when you use the specific blocks for it.

**Comment by djc** [comment]
(I also think custom escapers are a niche, and so making that feature a little harder to use to trade off run-time cost for potentially many more users is a trade-off I'd be willing to make.)

**Comment by cipriancraciun** [comment]
OK, let me ask this another way:  what is the cost to letting `{% space %}` and friends in the implementation?
* I'm reusing the `Node::Raw`;  (previously I've added a new `Node::Insert`, but now I see it's not necessary);  (basically I haven't touched the generator once I remove my `Node::Insert`;)
* I've added an `space_node` function for the parser;  (which is added at the end, thus the parser speed is not affected in normal conditions;)
* I'm using `space`, `newline` and `tab` blocks which aren't likely to be used for other purposes;

On the plus side:
* it covers some corner-cases when used with `stripspace`;
* it is compatible with those that switch from Go's `quicktemplate`;

**Comment by djc** [comment]
In the case of `space` and friends I'm not worried about the implementation costs but about the costs of understanding for template readers what's going on. (BTW, you didn't answer the question how an escaper that wants to escape whitespace would deal with the new blocks.)

**Comment by cipriancraciun** [comment]
> In the case of `space` and friends I'm not worried about the implementation costs but about the costs of understanding for template readers what's going on.

I've tried to address that in the documentation by stating:
~~~~
+//! Within the `stripspace` blocks (and not only) one could use `{% space %}`,
+//! `{% tab %}` and `{% newline %}` to force spaces before and after other blocks.
+//!
~~~~

----

> how an escaper that wants to escape whitespace would deal with the new blocks.

These new blocks are never passed through an escaper (which is by design), because they behave just as if a space was in the template in that exact same spot, namely the following two templates are identical in terms of generated code and render behaviour regardless of the escaper: `a b` and `a {%- space -%} b`. Moreover even today the escaper can't escape literal text that is not part of an `{{ x }}` expression.

The `{% space %}` is useful in the context of `{% stripspaces %}` as such:
~~~~
{% stripspace %}
    a
    {% space %}
    {% if condition %}
        b
    {% else %}
        c
    {% endif %}
    {% space %}
    d
{% endstripspace %}
~~~~

Where it would render as either `a b d` or `a c d` depending on the condition.  Without the two `{% space %}` blocks it would have always expanded as `abd` or `acd`.

Also that template is equivalent of:
~~~~
a {% if condition -%} b {%- else -%} c {%- endif %} d
~~~~

Although the `stripspace` is useful when the inner template is bigger, like plain-text report generation, etc.
